### ISSUE TITLE
No way to change external IP address of node

### ISSUE THREAD

**Issue by josmo** [rancher/rancher#14102]
If you shut down a node and restart it where the external IP address is different (ie. shutdown an instance in aws and restart it with a different ip)

Expected: Should have a way to set the new ip in rancher

Actual: uses the old ip and causes communication problems that makes the cluster unusable.

---
| Useful | Info |
| :-- | :-- |
|Versions|Rancher `v2.0.2` UI: `v2.0.48` |
|Access|`` `admin`|
|Route|`authenticated.cluster.nodes.index`|

**Comment by slawekww** [comment]
Got the same issue. Once my server has re-assigned IP by DHCP, K8S nodes cannot be recognized by Rancher. Solution is to register nodes as DNS name, not IPs.
See how many TLS certs in /etc/kubernetes are signed off using 'registered address'.

**Comment by josmo** [comment]
@slawekww do you mean using the dns for the --address and --internal-address?

**Comment by praveenperera** [comment]
I also have the same issue, the AWS nodes IPs changed and now I am unable to access them through rancher. Is there anyway to change these manually? I checked the API and wasn't able to.

I even tried editing the hosts file in the docker container to redirect the old ip to the new one. Had no luck.

**Comment by josmo** [comment]
Sadly I had to rebuild my cluster and start from scratch :/

**Comment by praveenperera** [comment]
@josmo were you able to launch the new cluster using elastic ips?

**Comment by francisco-sanchez-molina** [comment]
In version 1.6 you can launch this:
`sudo docker run -d -e CATTLE_AGENT_IP=<PRIVATE_IP> -v /var/run/docker....`

In 2.0 I can not find how to change node ip

**Comment by sirwolfgang** [comment]
I just tried do a reimport of a node like this:
```bash
sudo docker run -d --privileged --restart=unless-stopped --net=host -v /etc/kubernetes:/etc/kubernetes -v /var/run:/var/run rancher/rancher-agent:v2.0.4 --server https://rancher.aquilacashflow.com --token TOKEN --address IP_ADDRESS --worker
```

Doesnt seem to change anything.

**Comment by loganhz** [comment]
Hi @superseb , do you have any idea about this issue?

**Comment by aemneina** [comment]
You can rerun the host registration without deleting anything on the node, and the new IP should be detected/reflected after a couple of minutes.

**Comment by praveenperera** [comment]
How should that be done?

**Comment by aemneina** [comment]
Sorry I shot off my comment without understanding this was a 2.0 issue.

**Comment by josmo** [comment]
@praveenperera I think the eips would work in aws (I haven't tried it yet, but I imagine there should still be a different way of handling it if the ip changes) :)

**Comment by praveenperera** [comment]
@josmo To the best of my knowledge the only "problem" with EIPs is that rancher doesn't (can't?) create nodes with EIPs already attached. And when you attach a EIP that IP will be different than the IP the node had when it was created.

You would have to manually create the nodes, attach the EIPs then [create a custom cluster](https://rancher.com/docs/rancher/v2.x/en/cluster-provisioning/rke-clusters/custom-nodes/) in rancher.

**Comment by josmo** [comment]
@praveenperera totally makes sense. I hadn't thought about whether or not the metadata service can return the eip for an instance or the lifecycle of when it gets added :/

**Comment by sirwolfgang** [comment]
That is correct, you can not add EIPs until the server is provisioned, and thus after rancher set the IP address. It's also important to understand that EIPs is not a particularly good solution for this. The default limit is 5 EIPs per account.

**Comment by michalmikolajczyk** [comment]
that bug actually makes the cluster inaccessible for me with kubectl and workload shell ssh. Related to https://github.com/rancher/rancher/issues/12509

**Comment by loganhz** [comment]
Checked with @superseb 

Static IP is required for now. https://rancher.com/docs/rancher/v2.x/en/installation/requirements/ (Networking section)

Please follow the following instruction to recover it.

https://rancher.com/docs/rancher/v2.x/en/faq/technical/#the-ip-address-of-a-node-changed-how-can-i-recover

k8s is working on it though, so we probably support it in the future

**Comment by miclefebvre** [comment]
Same thing happen when ip change on vSphere

**Comment by appacea** [comment]
@loganhz thanks but those instructions don't really explain how to 'transpose' a node after the ip changed. It looks like I have to recreate my whole cluster (including workloads..etc).

This time around I want to use hostnames instead of IPs to avoid this problem in the future. Does anyone know if this is possible. When I tested, the cluster node seems to register both the hostname and the IP and it's not clear if that association can be updated manually if ever the IP changes.

**Comment by Akito13** [comment]
If the original request would be executed and even automatizable, that would be absolutely amazing. This would lift tons of pressure off certain projects...

**Comment by wsaca** [comment]
I'm using a node template with vSphere but if the IP is replaced the cluster stops working. Should this be resolved on the vsphere side or rancher should have the tools to recover from this?

**Comment by dipakparmar** [comment]
Same issue can't access the node after IP changed, the node uses internal IP and want to change to external IP. Any documentation on how to change manually?

**Comment by slawekww** [comment]
No, please add node with new IP and remove old node - it is only option I've used for my environments.

**Comment by olds463** [comment]
This renders the Cloud driver for AWS utterly useless then, no? There's no way to set an EIP even after cluster is created? Without using cloud driver, you lose ability to scale nodes up/down, use PVC EBS volumes and use ECR registries.  Cloud driver uses node templates to deploy, but if IP changes of nodes it completely breaks cluster. If you don't use EIP and amazon has fault that causes DHCP to get renewed your cluster will break if majority in same AZ/fault region.

There's no way around this?

**Comment by dlednik** [comment]
FYI:

I tried to update it via the API as I see `externalIpAddress` is an option in the JSON payload:
![image](https://user-images.githubusercontent.com/5610466/84980275-d8766700-b15b-11ea-95e1-1bc7b51205fb.png)

But unfortunately the new IP I sent with the payload is ignored.

**Comment by Nefelim4ag** [comment]
Just edit labels in K8s node by Kubectl, that will work.

**Comment by mateuszdrab** [comment]
I think I got into this issue. DHCP lease expired (presumably the node didn't renew it) and cluster was partially dead after the nodes rebooted.
What's worse, rancher allowed me to hit the scale down button enough time to bring the node count to 0 - removing all the nodes and hence destroying the cluster.
I've grabbed an etcd backup and I'm trying to figure out how to restore it into a new set of nodes - once those build. 
Currently they're stuck on `waiting for viable init node`

**Comment by Pluies** [comment]
Old ticket, but still very relevant, so posting for future use!

I ran into the same issue here: one node moved within a DC and received a new internal + external IP address, and I needed to update it.

My first attempt was to edit `/etc/rancher/rke2/config.yaml.d/50-rancher.yaml` on the node, which allows `rke2-agent` to restart, but the file was then overwritten with the old IPs and further restarts would break the node again.

In the end, I managed to solve it by editing the `RKEBootstrap` object associated to that machine on the management cluster.

Steps (on the management cluster):
1. Find the internal node id for the node you want to edit: `kubectl -n fleet-default get machines.cluster.x-k8s.io`. It should be of the form `custom-a1ef2c6ce90d`
2. Edit the RKEBootstrap object corresponding to this node: `kubectl -n fleet-default edit rkebootstrap custom-a1ef2c6ce90d`
  a. Update the `rke.cattle.io/address` annotation with the new external address
  b. Update the `rke.cattle.io/internal-address` annotation with the new internal address
3. Once this is done, Rancher will trigger a new plan, update the `/etc/rancher/rke2/config.yaml.d/50-rancher.yaml` config on the node with the new IPs, and everything is good again!

Our nodes are bare-metal nodes bootstrapped with `system-agent-install.sh`, so YMMV.
